Translatable: PropType for "translatable strings" instead of only plain strings

akirk opened the pull request:
This addresses warnings like
```
Warning: Failed prop type: Invalid prop `sectionTitle` of type `object` supplied to `SidebarNavigation`, expected `string`.
```

When the Community Translator is activated the return value of `translate()` is not guaranteed to be a `string` but can be a React component, thus a `PropTypes.string` check fails.

By introducing a new PropType that allows both a string and a component named ~`Translatable`~ `data`, we can still guarantee the correctness of the prop.

This also re-introduces support for `textOnly` (whose error message was useful to remove still existing textOnly uses) but we might be able to remove this as well.

ramonjd commented:
Great idea @akirk 

I rebased the branch with today's changes, and tested with some other components for example `SidebarBanner` and `NavTabs`. Swallows those prop warnings like a boss!

CircleCI checks are mostly eslint `Missing JSDoc` etc from my branch.

I wonder if we could create a PR out of this, and try to merge it into master on its own? Any way we can split up large PRs into small wins gets a 👍  from me.

akirk commented:
Right, I have rebased this onto master and changed the checked component to `data`: https://github.com/Automattic/wp-calypso/pull/21917/files#diff-e14997fc117d357458246233214e877fR15

Also, I have gone through some more of Calypso and identified further places to replace `PropTypes.string`.

ramonjd commented on client/lib/translator-jumpstart/index.js:
```diff
@@ -91,6 +91,11 @@ const communityTranslatorJumpstart = {
+			debug( 'respecting textOnly for string "' + originalFromPage + '"' );
+			return displayedTranslationFromPage;
+		}
+
```
Nice. 👍  For #21591 I'll need this to disable the translator for the translator ;)

akirk replied:
Really? I actually only used this to identify textOnly entries (using `localStorage.setItem('debug', 'calypso:community-translator')`) that I'd like to eliminate eventually. Having text wrapped in `<Translatable>` should still keep it usable.

ramonjd replied:
Ah, I see. I think there should be some way to circumvent wrapping a translation. I can only think of the CT dialog box as an example. Kinda like Translatable inception in that case :)

This seems too elaborate...

```
const closeText = this.props.translate( 'Close' );
<Button primary onClick={ this.closeDialog }>
   { closeText.toString() }
</Button>
```

ramonjd commented on client/components/translatable/proptype.js:
```diff
@@ -0,0 +1,62 @@
+/** @format */
+
+function translatableStringChecker( props, propName, componentName ) {
```
It is worth export this func and writing a test?

akirk replied:
Definitely. I'd be grateful for a little guidance! So how would I best approach this?

ramonjd replied:
I was half-way through a post, then I just though it would be easier to create a PR for this branch: #22031

While testing I noticed that we're letting through nil values (`null`, `undefined`) and `0`, while other numbers `2` will be validated. 

The tests I wrote are there to be tweaked :) It depends on how strict we want to be.

akirk replied:
I've updated code and tests to not allow numbers.
